### PR TITLE
refactor: salesforce test-cases with mocks

### PR DESCRIPTION
**Pull request by sanpj2292**
Signed-off-by: Sai Sankeerth <sanpj2292@github.com>

## Description of the change

We have implemented tests for `salesforce` in a way that is more extensible and common for `processor` & `router` transformation for all the destinations.
If this pattern looks good, we can plan to migrate to this kind of testing pattern for all future destinations

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: test/__tests__/transformation/proc/proc.test.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      });
+
+      inputData.forEach(async (input, index) => {
+        it(`Payload - ${index}`, async () => {
+          try {
+            const output = await process(input);
+            expect(output).toEqual(expectedData[index]);
+          } catch (error) {
+            expect(error.message).toEqual(expectedData[index].error);
+          }
```
Comment (saikumarrs) [inline test/__tests__/transformation/proc/proc.test.js:34]: This is a much better approach as the test case input dictates the expectations, and also, we'll be using the Jest framework components.
```suggestion
          if (expectedData[index].error) {
            expect(() => {
              await process(input);
              }).toThrow(expectedData[index].error);
          } else {
            const output = await process(input);
            expect(output).toEqual(expectedData[index]);
          }
```

**Review by saikumarrs — COMMENTED** [review]
LGTM, but @ujjwal-ab is already working on streamlining the test suite. You can pass on these learnings to him to avoid duplication of work.

For now, I'd recommend fixing the existing test suite to add the missing cases, if possible.

**Comment by sanpj2292** [comment]
> For now, I'd recommend fixing the existing test suite to add the missing cases, if possible.

Missing case which is causing the `Cannot read property 'access_token' of undefined` cannot be added without making changes to existing `axios.mock.js` & it is quite hard to do it.

Hence actually made complete U-turn & implement everything from scratch kind-of

File: test/__tests__/transformation/proc/proc.test.js (hunk 1 above)
Comment (sanpj2292) [reply test/__tests__/transformation/proc/proc.test.js:34]: Makes sense 👍🏼

**Comment by saikumarrs** [comment]
> > For now, I'd recommend fixing the existing test suite to add the missing cases, if possible.
> 
> Missing case which is causing the `Cannot read property 'access_token' of undefined` cannot be added without making changes to existing `axios.mock.js` & it is quite hard to do it.
> 
> Hence actually made complete U-turn & implement everything from scratch kind-of

Ok. @ujjwal-ab, let's account for this format in your investigation and come up with a final approach for the overall refactoring.

**Comment by ujjwal-ab** [comment]
> > > For now, I'd recommend fixing the existing test suite to add the missing cases, if possible.
> > 
> > 
> > Missing case which is causing the `Cannot read property 'access_token' of undefined` cannot be added without making changes to existing `axios.mock.js` & it is quite hard to do it.
> > Hence actually made complete U-turn & implement everything from scratch kind-of
> 
> Ok. @ujjwal-ab, let's account for this format in your investigation and come up with a final approach for the overall refactoring.

Yeah, this is kind of what we are thinking. But not sure, is it possible here to run proc/router tests for a single destination if we want to?

**Comment by sanpj2292** [comment]
> > > > For now, I'd recommend fixing the existing test suite to add the missing cases, if possible.
> > > 
> > > 
> > > Missing case which is causing the `Cannot read property 'access_token' of undefined` cannot be added without making changes to existing `axios.mock.js` & it is quite hard to do it.
> > > Hence actually made complete U-turn & implement everything from scratch kind-of
> > 
> > Ok. @ujjwal-ab, let's account for this format in your investigation and come up with a final approach for the overall refactoring.
> 
> Yeah, this is kind of what we are thinking. But not sure, is it possible here to run proc/router tests for a single destination if we want to?

You can run single test bybusing specific jest command using regex while executing jest command

**Comment by sanpj2292** [comment]
Implemented another way of mocking network requests such that we don't need to split the test-cases for processor & router transformations. The changes can be seen in [this commit](https://github.com/rudderlabs/rudder-transformer/pull/1740/commits/38dca1550bfbd7dae7319551e34b03d6bbba9a3a)

**Pull request closed without merging**